File: assets/code_examples/gme.py
import numpy as np
import pandas as pd
from scipy.optimize import minimize
from scipy.stats import chi2

class GMELogit:
    def __init__(self):
        self.beta = None
        self.vcov = None
        self.N = None
        self.K = None
        self.lnf = None
        self.lnf0 = None
        self.Sp = None
        self.S = None

    def f(self, x):
        """
        Compute f(x) with specified behavior:
        - f(x) = 0 if x is zero
        - f(x) = x * log(x) if x is not zero
        """
        return np.where(x == 0, 0, x * np.log(x))

    def fit(self, X, y):
        self.N, self.K = X.shape
        
        # Create error vector, Symmetric error support vector
        v1 = np.array([-1/np.sqrt(self.N), 0, 1/np.sqrt(self.N)])
        
        # Optimization
        initial_params = np.zeros(self.K)
        result_int = minimize(self._objective, initial_params, args=(X, y, v1), method='Nelder-Mead',
                          options = {'maxiter': 1000, 'disp': False})
        init = result_int.x
        result = minimize(self._objective, initial_params, args=(X, y, v1), method='BFGS',
                          options = {'maxiter': 1000, 'disp': False})
        
        self.beta = result.x
        self.vcov = result.hess_inv
        self.lnf = -result.fun
        self.lnf0 = self._objective(np.zeros_like(self.beta), X, y, v1)
        
        # Calculate normalized entropy
        P = self._calculate_probabilities(X)
        self.S = -np.sum(self.f(P) + self.f(1 - P))
        self.Sp = self.S / (self.N * np.log(2))
        
    def _objective(self, params, X, y, v1):
        # Ensure X is a numpy array
        if isinstance(X, pd.DataFrame):
            X = X.values

        # Ensure y is a numpy array
        if isinstance(y, pd.Series):
            y = y.values

        # Calculate predictions
        predictions = 1 / (1 + np.exp(-np.dot(X, params)))

        # Calculate log likelihood
        log_likelihood = -np.sum(y * np.log(predictions) + (1 - y) * np.log(1 - predictions))
        
        return log_likelihood

    def _calculate_probabilities(self, X):
        return 1 / (1 + np.exp(-X @ self.beta))

    def predict_proba(self, X):
        return self._calculate_probabilities(X)

    def summary(self):
        print("Generalized Maximum Entropy (Logit)")
        print(f"Number of obs: {self.N}")
        print(f"Degrees of freedom: {self.K}")
        print(f"Entropy for probs: {self.S:.1f}")
        print(f"Normalized entropy: {self.Sp:.4f}")
        
        # Calculate Entropy Ratio Statistic
        ERS = 2 * self.N * np.log(2) * (1 - self.Sp)
        p_value = chi2.sf(ERS, self.K)
        
        print(f"Ent. ratio stat: {ERS:.1f}")
        print(f"P Val for LR: {p_value:.4f}")
        print(f"Criterion F (log L) = {self.lnf:.4f}")
        print(f"Pseudo R2: {1 - self.Sp:.4f}")
        
        print("\nCoefficients:")
        for i, coef in enumerate(self.beta):
            std_err = np.sqrt(self.vcov[i, i])
            t_stat = coef / std_err
            p_value = 2 * (1 - chi2.cdf(t_stat**2, 1))
            print(f"X{i}: {coef:.4f} (std err: {std_err:.4f}, t-stat: {t_stat:.4f}, p-value: {p_value:.4f})")

# Usage example
if __name__ == "__main__":
    # Generate some example data
    np.random.seed(0)
    X = np.random.randn(1000, 5)
    true_beta = np.array([0.5, -0.3, 0.7, -0.2, 0.1, 0.6])  # including intercept
    y = (1 / (1 + np.exp(-X @ true_beta[:-1] - true_beta[-1])) > 0.5).astype(int)

    # Fit the model
    model = GMELogit()
    model.fit(X, y)

    # Print summary
    model.summary()

    # Make predictions
    X_new = np.random.randn(10, 5)
    predictions = model.predict_proba(X_new)
    print("\nPredictions for new data:")
    print(predictions)

File: assets/code_examples/gme.R
library(numDeriv)
library(stats)

GMELogit <- R6::R6Class("GMELogit",
                        public = list(
                          beta = NULL,
                          vcov = NULL,
                          N = NULL,
                          K = NULL,
                          lnf = NULL,
                          lnf0 = NULL,
                          Sp = NULL,
                          S = NULL,
                          
                          f = function(x) {
                            ifelse(x == 0, 0, x * log(x))
                          },
                          
                          fit = function(X, y) {
                            self$N <- nrow(X)
                            self$K <- ncol(X)
                            
                            v1 <- c(-1/sqrt(self$N), 0, 1/sqrt(self$N))
                            
                            initial_params <- rep(0, self$K)
                            result_int <- optim(initial_params, self$objective, X = X, y = y, v1 = v1, method = "Nelder-Mead",
                                                control = list(maxit = 1000))
                            init <- result_int$par
                            result <- optim(init, self$objective, X = X, y = y, v1 = v1, method = "BFGS",
                                            control = list(maxit = 1000), hessian = TRUE)
                            
                            self$beta <- result$par
                            self$vcov <- solve(result$hessian)
                            self$lnf <- -result$value
                            self$lnf0 <- self$objective(rep(0, self$K), X, y, v1)
                            
                            P <- self$calculate_probabilities(X)
                            self$S <- -sum(self$f(P) + self$f(1 - P))
                            self$Sp <- self$S / (self$N * log(2))
                          },
                          
                          objective = function(params, X, y, v1) {
                            predictions <- 1 / (1 + exp(-X %*% params))
                            -sum(y * log(predictions) + (1 - y) * log(1 - predictions))
                          },
                          
                          calculate_probabilities = function(X) {
                            1 / (1 + exp(-X %*% self$beta))
                          },
                          
                          predict_proba = function(X) {
                            self$calculate_probabilities(X)
                          },
                          
                          summary = function() {
                            cat("Generalized Maximum Entropy (Logit)\n")
                            cat("Number of obs:", self$N, "\n")
                            cat("Degrees of freedom:", self$K, "\n")
                            cat("Entropy for probs:", sprintf("%.1f", self$S), "\n")
                            cat("Normalized entropy:", sprintf("%.4f", self$Sp), "\n")
                            
                            ERS <- 2 * self$N * log(2) * (1 - self$Sp)
                            p_value <- pchisq(ERS, self$K, lower.tail = FALSE)
                            
                            cat("Ent. ratio stat:", sprintf("%.1f", ERS), "\n")
                            cat("P Val for LR:", sprintf("%.4f", p_value), "\n")
                            cat("Criterion F (log L) =", sprintf("%.4f", self$lnf), "\n")
                            cat("Pseudo R2:", sprintf("%.4f", 1 - self$Sp), "\n")
                            
                            cat("\nCoefficients:\n")
                            for (i in 1:length(self$beta)) {
                              coef <- self$beta[i]
                              std_err <- sqrt(self$vcov[i, i])
                              t_stat <- coef / std_err
                              p_value <- 2 * (1 - pchisq(t_stat^2, 1))
                              cat(sprintf("X%d: %.4f (std err: %.4f, t-stat: %.4f, p-value: %.4f)\n",
                                          i-1, coef, std_err, t_stat, p_value))
                            }
                          }
                        )
)

# Usage example
# set.seed(0)
# X <- matrix(rnorm(5000), nrow = 1000)
# true_beta <- c(0.5, -0.3, 0.7, -0.2, 0.1, 0.6)  # including intercept
# y <- as.integer((1 / (1 + exp(-X %*% true_beta[-length(true_beta)] - true_beta[length(true_beta)]))) > 0.5)
# 
# # Fit the model
# model <- GMELogit$new()
# model$fit(X, y)
# 
# # Print summary
# model$summary()

# Make predictions
# X_new <- matrix(rnorm(50), nrow = 10)
# predictions <- model$predict_proba(X_new)
# cat("\nPredictions for new data:\n")
# print(predictions)